File: scripts/textures.js
var fs = require('fs');
var config = require('../config');
var PNG = require('pngjs').PNG;

var numTextures = Object.keys(config.textures).length;
var width = 256.0;
var desiredHeight = 256.0 * numTextures;
var height = 256;
var p;

while (height < desiredHeight) {
    height = height<<1;
}

p = new PNG({width: width, height: height});

var yOffset = 0;
var out = {};
for (var value in config.textures) {
	var path = config.textures[value];
    var data = fs.readFileSync('../www' + path);
    var input = PNG.sync.read(data);
    console.log(path);

    //for (var i = 0; i < 32; i++) {
        var i = 0;
        PNG.bitblt(input, p, 0, 0, input.width, input.height, i * input.width, yOffset);
    //}

    var texcoordHeight = input.height / height;
    var texcoordWidth = input.width / width;
    var texcoordTop = 1.0 - (yOffset / height);
    var texcoordBottom = texcoordTop - texcoordHeight;

    out[value] = [texcoordBottom, texcoordTop, texcoordWidth];
    yOffset += input.height;
}

var options = { colorType: 6 };
var buffer = PNG.sync.write(p, options);
fs.writeFileSync('../www/textures.png', buffer);

fs.writeFileSync('../texture-offsets.js', 'module.exports=' + JSON.stringify(out));

File: scripts/textures.py
import json
import re
from PIL import Image

# pip3install pillow
# python3 textures.py

# read json file
with open('../config.mjs', 'r') as f:
    j = f.read()

    # remove comments and ESM stuff
    j = j.replace('export default ', '')
    j = re.sub(r"(// [^\n]+)", '', j)
    config = json.loads(j)


numTextures = len(config['textures'])

# lets try to include 3 of a texture vertically, so mipmapping is more accurate

width = 128
desiredHeight = width * numTextures * 3
height = width

# actual  height must be multiple of 2 for WebGL
while height < desiredHeight:
    height = height << 1
    
combined = Image.new('RGBA', (width, height), )

yOffset = 0
out = {}
for value in config['textures']:
    print(value)
    path = config['textures'][value]

    with Image.open('../www' + path) as image:
        if image.width != width:
            print('resizing %d to %d' % (image.width, width))
            image = image.resize( (width, width), resample=Image.NEAREST )
            
        i = 0
        while i < 3:
            combined.paste(image, (0, yOffset))

            if i == 1:
                texcoordTop = yOffset / height
                texcoordBottom = texcoordTop + ((width - 1) / height)
                out[value] = (texcoordBottom, texcoordTop, 1.0)

            yOffset += width
            i += 1
        yOffset += 2

with open('../texture-offsets.js', 'w') as f:
    a = 'module.exports=%s' % json.dumps(out)
    f.write(a)
    f.close()

flipped = combined.transpose(Image.FLIP_TOP_BOTTOM)
flipped.save('../www/textures.png', quality=100.0)
